### ISSUE TITLE
n.s.c.sse.ServerEventsClient memory leak and/or very frequent Hartbeats

### ISSUE THREAD

**Issue by kgoderis** [ServiceStack/ServiceStack.Java#28]
The C# documentation suggests that the default interval to send a SSE Heartbeat is 10 seconds (https://github.com/ServiceStack/ServiceStack/wiki/Server-Events). However, when registering a ServerEventsClient in Java, I observe *very* frequent Hartbeats being sent (I do not trigger them manually):

```
...
2017-07-04 23:46:35.451 [TRACE] [ing.atsadvanced.handler.PanelHandler] - DEBUG: [SSE-CLIENT] Heartbeat sent to: http://192.168.0.6:1337/event-heartbeat?id=AZtn8FX2ZnnGWKfAtnru
2017-07-04 23:46:35.452 [TRACE] [ing.atsadvanced.handler.PanelHandler] - DEBUG: [SSE-CLIENT] LastPulseAt: 23:46:35.452
2017-07-04 23:46:35.452 [TRACE] [ing.atsadvanced.handler.PanelHandler] - DEBUG: [SSE-CLIENT] OnHeartbeatReceived: (ServerEventHeartbeat) #12890 on #user1 ()
2017-07-04 23:46:35.483 [TRACE] [ing.atsadvanced.handler.PanelHandler] - DEBUG: [SSE-CLIENT] Prep for Heartbeat...
2017-07-04 23:46:35.483 [TRACE] [ing.atsadvanced.handler.PanelHandler] - DEBUG: [SSE-CLIENT] Sending Heartbeat...
2017-07-04 23:46:35.484 [TRACE] [ing.atsadvanced.handler.PanelHandler] - DEBUG: [SSE-CLIENT] Heartbeat sent to: http://192.168.0.6:1337/event-heartbeat?id=AZtn8FX2ZnnGWKfAtnru
2017-07-04 23:46:35.486 [TRACE] [ing.atsadvanced.handler.PanelHandler] - DEBUG: [SSE-CLIENT] LastPulseAt: 23:46:35.486
2017-07-04 23:46:35.486 [TRACE] [ing.atsadvanced.handler.PanelHandler] - DEBUG: [SSE-CLIENT] OnHeartbeatReceived: (ServerEventHeartbeat) #12891 on #user1 ()
2017-07-04 23:46:35.639 [TRACE] [ing.atsadvanced.handler.PanelHandler] - DEBUG: [SSE-CLIENT] Prep for Heartbeat...
2017-07-04 23:46:35.639 [TRACE] [ing.atsadvanced.handler.PanelHandler] - DEBUG: [SSE-CLIENT] Sending Heartbeat...
2017-07-04 23:46:35.640 [TRACE] [ing.atsadvanced.handler.PanelHandler] - DEBUG: [SSE-CLIENT] Heartbeat sent to: http://192.168.0.6:1337/event-heartbeat?id=AZtn8FX2ZnnGWKfAtnru
2017-07-04 23:46:35.641 [TRACE] [ing.atsadvanced.handler.PanelHandler] - DEBUG: [SSE-CLIENT] LastPulseAt: 23:46:35.641
2017-07-04 23:46:35.641 [TRACE] [ing.atsadvanced.handler.PanelHandler] - DEBUG: [SSE-CLIENT] OnHeartbeatReceived: (ServerEventHeartbeat) #12892 on #user1 ()
2017-07-04 23:46:35.723 [TRACE] [ing.atsadvanced.handler.PanelHandler] - DEBUG: [SSE-CLIENT] Prep for Heartbeat...
2017-07-04 23:46:35.724 [TRACE] [ing.atsadvanced.handler.PanelHandler] - DEBUG: [SSE-CLIENT] Sending Heartbeat...
2017-07-04 23:46:35.725 [TRACE] [ing.atsadvanced.handler.PanelHandler] - DEBUG: [SSE-CLIENT] Heartbeat sent to: http://192.168.0.6:1337/event-heartbeat?id=AZtn8FX2ZnnGWKfAtnru
2017-07-04 23:46:35.727 [TRACE] [ing.atsadvanced.handler.PanelHandler] - DEBUG: [SSE-CLIENT] LastPulseAt: 23:46:35.727
2017-07-04 23:46:35.727 [TRACE] [ing.atsadvanced.handler.PanelHandler] - DEBUG: [SSE-CLIENT] OnHeartbeatReceived: (ServerEventHeartbeat) #12893 on #user1 ()
2017-07-04 23:46:35.776 [TRACE] [ing.atsadvanced.handler.PanelHandler] - DEBUG: [SSE-CLIENT] Prep for Heartbeat...
2017-07-04 23:46:35.776 [TRACE] [ing.atsadvanced.handler.PanelHandler] - DEBUG: [SSE-CLIENT] Sending Heartbeat...
2017-07-04 23:46:35.777 [TRACE] [ing.atsadvanced.handler.PanelHandler] - DEBUG: [SSE-CLIENT] Heartbeat sent to: http://192.168.0.6:1337/event-heartbeat?id=AZtn8FX2ZnnGWKfAtnru
2017-07-04 23:46:35.778 [TRACE] [ing.atsadvanced.handler.PanelHandler] - DEBUG: [SSE-CLIENT] LastPulseAt: 23:46:35.778
2017-07-04 23:46:35.778 [TRACE] [ing.atsadvanced.handler.PanelHandler] - DEBUG: [SSE-CLIENT] OnHeartbeatReceived: (ServerEventHeartbeat) #12894 on #user1 ()
2017-07-04 23:46:35.972 [TRACE] [ing.atsadvanced.handler.PanelHandler] - DEBUG: [SSE-CLIENT] Prep for Heartbeat...
2017-07-04 23:46:35.972 [TRACE] [ing.atsadvanced.handler.PanelHandler] - DEBUG: [SSE-CLIENT] Sending Heartbeat...
2017-07-04 23:46:35.973 [TRACE] [ing.atsadvanced.handler.PanelHandler] - DEBUG: [SSE-CLIENT] Heartbeat sent to: http://192.168.0.6:1337/event-heartbeat?id=AZtn8FX2ZnnGWKfAtnru
2017-07-04 23:46:35.974 [TRACE] [ing.atsadvanced.handler.PanelHandler] - DEBUG: [SSE-CLIENT] LastPulseAt: 23:46:35.974
2017-07-04 23:46:35.975 [TRACE] [ing.atsadvanced.handler.PanelHandler] - DEBUG: [SSE-CLIENT] OnHeartbeatReceived: (ServerEventHeartbeat) #12895 on #user1 ()
...
```

When running for a long time, it tends to clog the Heap of the JVM, in my case almost allocating 5GB to the n.s.c.s.sec class, indicating that there must be some kind of memory leak in that class as well
[btw, unless I am mistaking here, when .close() is called on the SSE client, the Timer task used for the Heartbeat is never stopped/cancelled]

**Comment by kgoderis** [comment]
After some analysis after running the SSE client for a short while: 

_One instance of "java.util.TaskQueue" loaded by "<system class loader>" occupies 230,996,576 (37,04%) bytes. The instance is referenced by net.servicestack.client.sse.ServerEventsClient @ 0x5d87ade60 , loaded by "org.openhab.binding.atsadvanced". The memory is accumulated in one instance of "java.util.TimerTask[]" loaded by "<system class loader>"._

In the code I see that a iterative mechanism is put in place to reschedule the TimerTask for the heartbeat. Maybe this should be done with an Executor and Features that can be better managed (e.g. cancelled property etc)?

**Comment by mythz** [comment]
What type of App are you running this in? e.g. Android, Java Web or Java Desktop App? Are you doing anything unusual like starting/stopping multiple times or is this run in a long-lived App that's configured and started once then runs continuously?

What do you mean by short while? 1-3, 5-10 mins or longer?

**Comment by kgoderis** [comment]
> What type of App are you running this in? e.g. Android, Java Web or Java Desktop App? Are you doing anything unusual like starting/stopping multiple times or is this run in a long-lived App that's configured and started once then runs continuously?

This is a Java Desktop App (www.openhab.org &  www.eclipse.be/smarthome)

> 
> What do you mean by short while? 1-3, 5-10 mins or longer?

A few hours I would say.

**Comment by mythz** [comment]
I've changed it to use `ScheduledThreadPoolExecutor` in this commit: https://github.com/ServiceStack/ServiceStack.Java/commit/64c6d5c84af98360d7bca480640b38f4fb6c8607

This change is available in v1.0.32, can you upgrade to the latest version and let me know if it resolves the issue?

**Comment by kgoderis** [comment]
@mythz Can you push the jar to MavenCentral?

**Comment by mythz** [comment]
@kgoderis it was [published to Maven Central](https://search.maven.org/#artifactdetails%7Cnet.servicestack%7Cclient%7C1.0.32%7Cjar) just takes a little while to sync.

**Comment by kgoderis** [comment]
@mythz I have the code running now. The ServerEventConnect reports back that heartbeats will be sent every 10000ms, but my logs show something different though. It is less problematic than before, still a massive number of heartbeats:

```
2017-07-06 20:00:03.164 [TRACE] [ing.atsadvanced.handler.PanelHandler] - DEBUG: [SSE-CLIENT] LastPulseAt: 20:00:03.164
2017-07-06 20:00:03.166 [TRACE] [ing.atsadvanced.handler.PanelHandler] - DEBUG: [SSE-CLIENT] LastPulseAt: 20:00:03.166
2017-07-06 20:00:03.177 [TRACE] [ing.atsadvanced.handler.PanelHandler] - DEBUG: [SSE-CLIENT] LastPulseAt: 20:00:03.177
2017-07-06 20:00:03.337 [TRACE] [ing.atsadvanced.handler.PanelHandler] - DEBUG: [SSE-CLIENT] LastPulseAt: 20:00:03.337
2017-07-06 20:00:03.492 [TRACE] [ing.atsadvanced.handler.PanelHandler] - DEBUG: [SSE-CLIENT] LastPulseAt: 20:00:03.492
2017-07-06 20:00:03.683 [TRACE] [ing.atsadvanced.handler.PanelHandler] - DEBUG: [SSE-CLIENT] LastPulseAt: 20:00:03.683
2017-07-06 20:00:03.849 [TRACE] [ing.atsadvanced.handler.PanelHandler] - DEBUG: [SSE-CLIENT] LastPulseAt: 20:00:03.849
2017-07-06 20:00:03.990 [TRACE] [ing.atsadvanced.handler.PanelHandler] - DEBUG: [SSE-CLIENT] LastPulseAt: 20:00:03.990
2017-07-06 20:00:07.570 [TRACE] [ing.atsadvanced.handler.PanelHandler] - DEBUG: [SSE-CLIENT] LastPulseAt: 20:00:07.570
2017-07-06 20:00:13.166 [TRACE] [ing.atsadvanced.handler.PanelHandler] - DEBUG: [SSE-CLIENT] LastPulseAt: 20:00:13.166
2017-07-06 20:00:13.167 [TRACE] [ing.atsadvanced.handler.PanelHandler] - DEBUG: [SSE-CLIENT] LastPulseAt: 20:00:13.167
2017-07-06 20:00:13.179 [TRACE] [ing.atsadvanced.handler.PanelHandler] - DEBUG: [SSE-CLIENT] LastPulseAt: 20:00:13.179
2017-07-06 20:00:13.339 [TRACE] [ing.atsadvanced.handler.PanelHandler] - DEBUG: [SSE-CLIENT] LastPulseAt: 20:00:13.339
2017-07-06 20:00:13.493 [TRACE] [ing.atsadvanced.handler.PanelHandler] - DEBUG: [SSE-CLIENT] LastPulseAt: 20:00:13.493
2017-07-06 20:00:13.684 [TRACE] [ing.atsadvanced.handler.PanelHandler] - DEBUG: [SSE-CLIENT] LastPulseAt: 20:00:13.684
2017-07-06 20:00:13.851 [TRACE] [ing.atsadvanced.handler.PanelHandler] - DEBUG: [SSE-CLIENT] LastPulseAt: 20:00:13.851
2017-07-06 20:00:13.992 [TRACE] [ing.atsadvanced.handler.PanelHandler] - DEBUG: [SSE-CLIENT] LastPulseAt: 20:00:13.992
2017-07-06 20:00:17.571 [TRACE] [ing.atsadvanced.handler.PanelHandler] - DEBUG: [SSE-CLIENT] LastPulseAt: 20:00:17.571
2017-07-06 20:00:23.167 [TRACE] [ing.atsadvanced.handler.PanelHandler] - DEBUG: [SSE-CLIENT] LastPulseAt: 20:00:23.167
2017-07-06 20:00:23.168 [TRACE] [ing.atsadvanced.handler.PanelHandler] - DEBUG: [SSE-CLIENT] LastPulseAt: 20:00:23.168
2017-07-06 20:00:23.179 [TRACE] [ing.atsadvanced.handler.PanelHandler] - DEBUG: [SSE-CLIENT] LastPulseAt: 20:00:23.179
2017-07-06 20:00:23.340 [TRACE] [ing.atsadvanced.handler.PanelHandler] - DEBUG: [SSE-CLIENT] LastPulseAt: 20:00:23.340
2017-07-06 20:00:23.495 [TRACE] [ing.atsadvanced.handler.PanelHandler] - DEBUG: [SSE-CLIENT] LastPulseAt: 20:00:23.495
2017-07-06 20:00:23.686 [TRACE] [ing.atsadvanced.handler.PanelHandler] - DEBUG: [SSE-CLIENT] LastPulseAt: 20:00:23.686
2017-07-06 20:00:23.852 [TRACE] [ing.atsadvanced.handler.PanelHandler] - DEBUG: [SSE-CLIENT] LastPulseAt: 20:00:23.852
2017-07-06 20:00:23.993 [TRACE] [ing.atsadvanced.handler.PanelHandler] - DEBUG: [SSE-CLIENT] LastPulseAt: 20:00:23.993
2017-07-06 20:00:24.158 [TRACE] [ing.atsadvanced.handler.PanelHandler] - DEBUG: [SSE-CLIENT] LastPulseAt: 20:00:24.158
2017-07-06 20:00:27.573 [TRACE] [ing.atsadvanced.handler.PanelHandler] - DEBUG: [SSE-CLIENT] LastPulseAt: 20:00:27.573
```

**Comment by kgoderis** [comment]
@mythz if you look at the logs, it tends to fire in batches every 10000 ms. It is triggered too much at the server side?

**Comment by kgoderis** [comment]
@mythz FYI, after running for 30 or so minutes, I now have about 6-7 heartbeats per second. I suspect there is something wrong at the server side (C# code, with default settings for the SSE plugin)

**Comment by mythz** [comment]
It's not the server, heartbeats are initiated from the client, the server holds the default config that instructs the client how frequently to send to heartbeats in `ServerEventsFeature.HeartbeatInterval` which defaults to 10 seconds whilst `IdleTimeout` tells the server how long the client can lapse sending a heartbeat before it considers the connection failed (default 30 secs).

I'm not sure why you're seeing frequent heartbeats, do you have multiple ServerEventsClient connected or doing something else unusual?

**Comment by kgoderis** [comment]
There is only a single SEC running. However, every 30 seconds, whilst polling the device running the SSE server, I initiate a manual Heartbeat(). 

Looking back at the code you changed, it seems to be quite expensive to initiate an Executor at each hartbeat. I would rather have a single one, schedule a job, and keep a reference to the ScheduledFeature instead, then keep track if the feature was executed. In case it is not, you could forego scheduling a new heartbeat.

**Comment by mythz** [comment]
It is reusing the same `ScheduledThreadPoolExecutor` instance, it just re-schedules a new one-shot callback after it receives a response from the last heartbeat: https://github.com/ServiceStack/ServiceStack.Java/blob/master/src/AndroidClient/client/src/main/java/net/servicestack/client/sse/ServerEventsClient.java#L573-L581

The multiple heartbeats could be causing the issue though, I'll look into prevent multiple calls to `startNewHeartbeat()` to only let one heartbeat request through.

**Comment by mythz** [comment]
It's avoiding a periodic scheduler because that is what could queue requests if the server starts getting overloaded and starts taking longer to respond to heartbeats which is why we only queue the next scheduler after we get a response back from the server for the last heartbeat.

I'd prefer throttling the current heartbeats then using a periodic scheduler.

**Comment by mythz** [comment]
> There is only a single SEC running. However, every 30 seconds, whilst polling the device running the SSE server, I initiate a manual Heartbeat(). That in itself does not explain the cascade (I think) 

This could definitely be the problem, each time you run `Heartbeat()` manually it accumulates new heartbeats schedule loops. Please try without initiating any new manual `Heartbeat()` calls. 

I've just tried [leaving it running for an hour](https://github.com/ServiceStack/ServiceStack.Java/commit/98c157ef3de34748556805ac501c025fcd1a2d51) and heartbeats are working as expected with the last series of heartbeats happening once in every 10 seconds.

If the issue persists, please include your code used to initialize and start the ServerEventsClient as well as any other manual code that touches the `ServerEventsClient` after it's running. (after removing any Heartbeat() from your code).

**Comment by kgoderis** [comment]
Removing that manual Heartbeat() solves it, but in that case I think you should remove the heartbeat() API all together, since the resulting behaviour is not what is expected at all

btw, scheduleWithFixedDelay() is not periodic scheduling at all. A new job will be only executed when the previous one has finished. (e.g it is not like scheduleAtFixedRate())

**Comment by mythz** [comment]
Great, glad its resolved.

Cool it sounds like `scheduleWithFixedDelay()` does what we want, modified to use it in [this commit](https://github.com/ServiceStack/ServiceStack.Java/commit/c95a6772afe397c4dba8990a2499103334dba687) which should also make the `Heartbeat()` API safe now.

This change is available from v1.0.33 that I've just deployed to bintray, it takes a little while to sync to MavenCentral.
